server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java

```diff
@@ -143,6 +143,12 @@ private static boolean hasUsefulTimestampField(IndexMetadata indexMetadata) {
         if (indexMetadata == null) {
             return false;
         }
+
+        if (indexMetadata.getTimeSeriesTimestampRange() != null) {
```

Comment:
After writing more tests and actually debugging the code, I noticed that I needed to add this if statement here. Otherwise `@timestamp` range queries against tsdb indices always execute can_match phase against nodes contain shards instead of resolving wether the query resolves to `MatchNoneQueryBuilder` on coordinating node.

I have no idea how to write a good test for this, that checks that shards are skipped because on the coordinating node we figured out that a query resolves to a `MatchNoneQueryBuilder`. 

The point of this PR, is that we should figure out on coordinating if a query has a range query against `@timestamp` field that doesn't match with the the start_time and end_time of indices. I can't check whether can_match search phase skipped shards or the query rewrite on the coordinating node. I would like to verify this,.

Reply:
Only `@timestamp` field of indices that are listed in `TimestampFieldMapperService` can do the rewrite on coordinating node if understand the code correctly.

Reply:
Ah! That makes sense.

This is the kind of thing I'd test with a yaml test if possible. It's sneaky enough that I'd want an "outside" test rather than a unit test. And it's important for performance.

Reply:
Hmmmm. Adding this means we're going to be capturing the mapper for the timestamp of every tsdb index on every node. I wonder if in a follow up we can make these singletons....

Reply:
> Hmmmm. Adding this means we're going to be capturing the mapper for the timestamp of every tsdb index on every node. I wonder if in a follow up we can make these singletons....

Yes. I think we can do that if the timestamp field mapping between data streams are configured in the same way. Caching the same timestamp field mapper is silly.

Reply:
> Is it available in the profiler?

I don't think we record this anywhere in `CanMatchPreFilter`. Would be great if we keep track of can match request like we do for query and fetch request in `SearchStats`. Then we can assert in tests that can_match counts haven't increased after doing a search.